field/src/extension/binomial_extension.rs

```diff
@@ -0,0 +1,368 @@
+    type Output = Self;
+
+    #[inline]
+    fn add(self, rhs: Self) -> Self {
+        let mut res = self.0;
+        for (r, &rhs_val) in res.iter_mut().zip(&rhs.0) {
+            *r += rhs_val;
+        }
+        Self(res)
+    }
+}
+
+impl<F: BinomiallyExtendable<D>, const D: usize> Add<F> for BinomialExtensionField<F, D> {
+    type Output = Self;
+
+    #[inline]
+    fn add(self, rhs: F) -> Self {
+        let mut res = self.0;
+        res[0] += rhs;
+        Self(res)
+    }
+}
+
+impl<F: BinomiallyExtendable<D>, const D: usize> AddAssign for BinomialExtensionField<F, D> {
+    fn add_assign(&mut self, rhs: Self) {
+        *self = *self + rhs;
+    }
+}
+
+impl<F: BinomiallyExtendable<D>, const D: usize> AddAssign<F> for BinomialExtensionField<F, D> {
+    fn add_assign(&mut self, rhs: F) {
+        *self = *self + rhs;
+    }
+}
+
+impl<F: BinomiallyExtendable<D>, const D: usize> Sum for BinomialExtensionField<F, D> {
+    fn sum<I: Iterator<Item = Self>>(iter: I) -> Self {
+        iter.fold(Self::ZERO, |acc, x| acc + x)
+    }
+}
+
+impl<F: BinomiallyExtendable<D>, const D: usize> Sub for BinomialExtensionField<F, D> {
+    type Output = Self;
+
+    #[inline]
+    fn sub(self, rhs: Self) -> Self {
+        let mut res = self.0;
+        for (r, &rhs_val) in res.iter_mut().zip(&rhs.0) {
+            *r -= rhs_val;
+        }
+        Self(res)
+    }
+}
+
+impl<F: BinomiallyExtendable<D>, const D: usize> Sub<F> for BinomialExtensionField<F, D> {
+    type Output = Self;
+
+    #[inline]
+    fn sub(self, rhs: F) -> Self {
+        let mut res = self.0;
+        res[0] -= rhs;
+        Self(res)
+    }
+}
+
+impl<F: BinomiallyExtendable<D>, const D: usize> SubAssign for BinomialExtensionField<F, D> {
+    #[inline]
+    fn sub_assign(&mut self, rhs: Self) {
+        *self = *self - rhs;
+    }
+}
+
+impl<F: BinomiallyExtendable<D>, const D: usize> SubAssign<F> for BinomialExtensionField<F, D> {
+    #[inline]
+    fn sub_assign(&mut self, rhs: F) {
+        *self = *self - rhs;
+    }
+}
+
+impl<F: BinomiallyExtendable<D>, const D: usize> Mul for BinomialExtensionField<F, D> {
+    type Output = Self;
+
+    #[inline]
+    fn mul(self, rhs: Self) -> Self {
+        let Self(a) = self;
+        let Self(b) = rhs;
+
+        let mut res = [F::ZERO; D];
+
+        for i in 0..D {
+            for j in 0..D {
+                if i + j >= D {
+                    res[i + j - D] += F::W * a[i] * b[j];
+                } else {
+                    res[i + j] += a[i] * b[j];
+                }
+            }
+        }
+
+        Self(res)
+    }
+}
+
+impl<F: BinomiallyExtendable<D>, const D: usize> Mul<F> for BinomialExtensionField<F, D> {
+    type Output = Self;
+
+    #[inline]
+    fn mul(self, rhs: F) -> Self {
+        Self(
+            self.0
+                .into_iter()
+                .map(|x| x * rhs)
+                .collect::<Vec<F>>()
+                .try_into()
+                .unwrap(),
+        )
```

Comment:
As with `neg()` above, this can be simplified:
```suggestion
            self.0.map(|x| x * rhs))
```
Having simplified these two statements, the `alloc::vec::Vec` import can also be removed.

Reply:
updated, thx.